packages/dht/src/dht/PeerManager.ts

```diff
@@ -0,0 +1,320 @@
+/* eslint-disable @typescript-eslint/parameter-properties, class-methods-use-this */
+
+import KBucket from 'k-bucket'
+import { PeerDescriptor } from '../proto/packages/dht/protos/DhtRpc'
+import { RemoteDhtNode } from './RemoteDhtNode'
+import { IPeerManager } from './IPeerManager'
+import { SortedContactList } from './contact/SortedContactList'
+import { RandomContactList } from './contact/RandomContactList'
+import { PeerID, PeerIDKey } from '../helpers/PeerID'
+import { Logger } from '@streamr/utils'
+import { keyFromPeerDescriptor, peerIdFromPeerDescriptor } from '../helpers/peerIdFromPeerDescriptor'
+import { ConnectionManager } from '../connection/ConnectionManager'
+import EventEmitter from 'eventemitter3'
+import { DisconnectionType } from '../transport/ITransport'
+
+const logger = new Logger(module)
+
+export interface PeerManagerConfig {
+    numberOfNodesPerKBucket: number
+    maxNeighborListSize: number
+    getClosestContactsLimit: number
+    ownPeerId: PeerID
+    connectionManager: ConnectionManager
+    isLayer0: boolean
+    createRemoteDhtNode: (peerDescriptor: PeerDescriptor) => RemoteDhtNode
+}
+
+export interface PeerManagerEvents {
+    newContact: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    contactRemoved: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    newKbucketContact: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    kbucketContactRemoved: (peerDescriptor: PeerDescriptor) => void
+    newOpenInternetContact: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    openInternetContactRemoved: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    newRandomContact: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    randomContactRemoved: (peerDescriptor: PeerDescriptor, closestPeers: PeerDescriptor[]) => void
+    kBucketEmpty: () => void
+}
+
+interface IPeerManagerTest {
+    getNeighborList: () => SortedContactList<RemoteDhtNode>
+    getKBucketSize: () => number
+    getKBucketPeers: () => PeerDescriptor[]
+}
+
+export class PeerManager extends EventEmitter<PeerManagerEvents> implements IPeerManager {
+    private _jee: number = 0
+    private bucket?: KBucket<RemoteDhtNode>
+    private neighborList?: SortedContactList<RemoteDhtNode>
+    private openInternetPeers?: SortedContactList<RemoteDhtNode>
+    private randomPeers?: RandomContactList<RemoteDhtNode>
```

Comment:
All these fields are initialized in the constructor (as it calls `initKBuckets()`), and won't be `undefined`.
```suggestion
    private bucket: KBucket<RemoteDhtNode>
    private neighborList: SortedContactList<RemoteDhtNode>
    private openInternetPeers: SortedContactList<RemoteDhtNode>
    private randomPeers: RandomContactList<RemoteDhtNode>
```